File: solutions/python/musicstreamingservice/music_library.py
class MusicLibrary:
    _instance = None

    def __init__(self):
        self.songs = {}
        self.albums = {}
        self.artists = {}

    @classmethod
    def get_instance(cls):
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    def add_song(self, song):
        self.songs[song.get_id()] = song

    def add_album(self, album):
        self.albums[album.get_id()] = album
        for song in album.get_songs():
            self.add_song(song)

    def add_artist(self, artist):
        self.artists[artist.get_id()] = artist
        for album in artist.get_albums():
            self.add_album(album)

    def get_song(self, song_id):
        return self.songs.get(song_id)

    def get_album(self, album_id):
        return self.albums.get(album_id)

    def get_artist(self, artist_id):
        return self.artists.get(artist_id)

    def search_songs(self, query):
        matching_songs = []
        for song in self.songs.values():
            if query in song.get_title() or query in song.get_artist() or query in song.get_album():
                matching_songs.append(song)
        return matching_songs

File: solutions/java/src/musicstreamingservice/MusicLibrary.java
package musicstreamingservice;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class MusicLibrary {
    private static MusicLibrary instance;
    private final Map<String, Song> songs;
    private final Map<String, Album> albums;
    private final Map<String, Artist> artists;

    private MusicLibrary() {
        songs = new ConcurrentHashMap<>();
        albums = new ConcurrentHashMap<>();
        artists = new ConcurrentHashMap<>();
    }

    public static synchronized MusicLibrary getInstance() {
        if (instance == null) {
            instance = new MusicLibrary();
        }
        return instance;
    }

    public void addSong(Song song) {
        songs.put(song.getId(), song);
    }

    public void addAlbum(Album album) {
        albums.put(album.getId(), album);
        for (Song song : album.getSongs()) {
            addSong(song);
        }
    }

    public void addArtist(Artist artist) {
        artists.put(artist.getId(), artist);
        for (Album album : artist.getAlbums()) {
            addAlbum(album);
        }
    }

    public Song getSong(String songId) {
        return songs.get(songId);
    }

    public List<Song> getSongs() {
        return songs.values().stream().toList();
    }

    public Album getAlbum(String albumId) {
        return albums.get(albumId);
    }

    public Artist getArtist(String artistId) {
        return artists.get(artistId);
    }
}
